File: internal/cmd/waves.go
package cmd

import (
	"fmt"

	"github.com/0xkhdr/specd/internal/cli"
	"github.com/0xkhdr/specd/internal/core"
)

// RunWaves implements `specd waves`: it loads the spec's task DAG and prints
// the wave-by-wave task graph, the critical path, and any blockers, as JSON
// (--json) or as the human-readable wave graph.
func RunWaves(args cli.Args) int {
	root, slug, code, ok := requireRootAndSlug(args, "usage: specd waves <slug> [--json]")
	if !ok {
		return code
	}
	loaded, err := core.LoadSpec(root, slug)
	if err != nil {
		return specdExit(err)
	}
	state := loaded.State

	if args.Bool("json") {
		tasks := core.DagTasksFromState(state)
		type waveJSON struct {
			Wave  int `json:"wave"`
			Tasks []struct {
				ID      string          `json:"id"`
				Status  core.TaskStatus `json:"status"`
				Depends []string        `json:"depends"`
			} `json:"tasks"`
		}
		waves := core.GroupWaves(tasks)
		wout := make([]waveJSON, len(waves))
		for i, w := range waves {
			wout[i].Wave = w.Wave
			for _, t := range w.Tasks {
				depends := t.Depends
				if depends == nil {
					depends = []string{}
				}
				wout[i].Tasks = append(wout[i].Tasks, struct {
					ID      string          `json:"id"`
					Status  core.TaskStatus `json:"status"`
					Depends []string        `json:"depends"`
				}{t.ID, t.Status, depends})
			}
		}
		critical := core.CriticalPath(tasks)
		if critical == nil {
			critical = []string{}
		}
		blockers := state.Blockers
		if blockers == nil {
			blockers = []core.Blocker{}
		}
		out := map[string]interface{}{
			"waves":        wout,
			"criticalPath": critical,
			"blockers":     blockers,
		}
		if err := core.PrintJSON(out); err != nil {
			return specdExit(err)
		}
		return core.ExitOK
	}

	fmt.Println(core.WaveGraph(state))
	if len(state.Blockers) > 0 {
		fmt.Println("\nBlockers gating downstream waves:")
		for _, b := range state.Blockers {
			fmt.Printf("  ⚠ %s: %s\n", b.Task, b.Reason)
		}
	}
	return core.ExitOK
}

File: src/commands/waves.ts
// `specd waves <slug> [--json]` — the wave DAG, critical path, blockers (SPEC §5.2, §7.3.3).
import { requireSpecdRoot } from "../core/paths.js";
import { loadSpec } from "../core/specFiles.js";
import { usageError } from "../core/exit.js";
import { criticalPath, groupWaves } from "../core/dag.js";
import { dagTasks, waveGraph } from "../core/render.js";
import type { Args } from "../cli.js";

export function run(args: Args): number {
  const root = requireSpecdRoot();
  const slug = args.pos[0];
  if (!slug) throw usageError("usage: specd waves <slug> [--json]");
  const { state } = loadSpec(root, slug);

  if (args.flags.json === true) {
    const tasks = dagTasks(state);
    const waves = groupWaves(tasks).map((r) => ({
      wave: r.wave,
      tasks: r.tasks.map((t) => ({ id: t.id, status: t.status, depends: t.depends })),
    }));
    console.log(JSON.stringify({
      waves,
      criticalPath: criticalPath(tasks),
      blockers: state.blockers,
    }, null, 2));
    return 0;
  }

  console.log(waveGraph(state));
  if (state.blockers.length) {
    console.log("\nBlockers gating downstream waves:");
    for (const b of state.blockers) console.log(`  ⚠ ${b.task}: ${b.reason}`);
  }
  return 0;
}
